web3_proxy/src/frontend/users/payment_stripe.rs

```diff
@@ -0,0 +1,236 @@
+
+    // Filter by user ...
+    let receipts = stripe_increase_balance_receipt::Entity::find()
+        .filter(increase_on_chain_balance_receipt::Column::DepositToUserId.eq(user.id))
+        .all(db_replica.as_ref())
+        .await?;
+
+    // Return the response, all except the user ...
+    let response = json!({
+        "user": Address::from_slice(&user.address),
+        "deposits": receipts,
+    });
+
+    Ok(Json(response).into_response())
+}
+
+/// `POST /user/balance/stripe/` -- Process a stripe transaction;
+/// this endpoint is called from the webhook with the user_id parameter in the request
+#[debug_handler]
+pub async fn user_balance_stripe_post(
+    Extension(app): Extension<Arc<Web3ProxyApp>>,
+    ip: InsecureClientIp,
+    headers: HeaderMap,
+    Path(mut params): Path<HashMap<String, String>>,
+    bearer: Option<TypedHeader<Authorization<Bearer>>>,
+) -> Web3ProxyResponse {
+    // rate limit by bearer token **OR** IP address
+    let (authorization, _semaphore) = if let Some(TypedHeader(Authorization(bearer))) = bearer {
+        let (_, semaphore) = app.bearer_is_authorized(bearer).await?;
+
+        // TODO: is handling this as internal fine?
+        let authorization = Web3ProxyAuthorization::internal(app.db_conn())?;
+
+        (authorization, Some(semaphore))
+    } else {
+        let InsecureClientIp(ip) = ip;
+        let authorization = login_is_authorized(&app, ip).await?;
+        (authorization, None)
+    };
+
+    let recipient_user_id: u64 = params
+        .remove("user_id")
+        .ok_or(Web3ProxyError::BadRouting)?
+        .parse()
+        .or(Err(Web3ProxyError::ParseAddressError))?;
+
+    // Get the payload, and the header
+    let payload = params.remove("data").ok_or(Web3ProxyError::BadRequest(
+        "You have not provided a 'data' for the Stripe payload".into(),
+    ))?;
+
+    // TODO Get this from the header
+    let signature = headers
+        .get("STRIPE_SIGNATURE")
+        .ok_or(Web3ProxyError::BadRequest(
+            "You have not provided a 'STRIPE_SIGNATURE' for the Stripe payload".into(),
+        ))?
+        .to_str()
+        .web3_context("Could not parse stripe signature as byte-string")?;
+
+    // Now parse the payload and signature
+    // TODO: Move env variable elsewhere
+    let event = Webhook::construct_event(
+        &payload,
+        &signature,
+        app.config
+            .stripe_api_key
+            .clone()
+            .web3_context("Stripe API key not found in config!")?
+            .as_str(),
+    )
+    .context(Web3ProxyError::BadRequest(
+        "Could not parse the stripe webhook request!".into(),
+    ))?;
+
+    let intent = match event.data.object {
+        stripe::EventObject::PaymentIntent(intent) => intent,
+        _ => return Ok("Received irrelevant webhook".into_response()),
+    };
+    debug!("Found PaymentIntent Event: {:?}", intent);
+
+    if intent.status.as_str() != "succeeded" {
```

Comment:
I feel like we probably want to do something more if it's not succeeded. Like, what if they post to tell us there was a chargeback?

Reply:
![Screenshot 2023-06-26 at 20 03 18](https://github.com/llamanodes/web3-proxy/assets/8946130/bdc4a239-d7d2-4a0a-a065-eae5c5d653bc)

which ones do you want to cover? also as I understand, this endpoint responds to the webhook, not the frontend, so not sure how we would best communicate it with the frontend

Reply:
What do we need to communicate to the frontend? They can just poll the user's balance endpoint, right?

Reply:
yes agree. i'll leave it at this then